scripts/setup-static-build.sh

```diff
@@ -0,0 +1,62 @@
+#!/bin/sh
+set -ue
+
+LC_ALL=C
+
+cd $(dirname "$0")/..
+
+## Static linking configuration ##
+
+# The linked C libraries list may need updating on changes to the dependencies.
+#
+# To get the correct list for manual linking, the simplest way is to set the
+# flags to `-verbose`, while on the normal `autolink` mode, then extract them
+# from the gcc command-line.
+
+case $(uname -s) in
+    Linux)
+        case $(. /etc/os-release && echo $ID) in
+            alpine)
+                COMMON_LIBS="camlstr base_stubs ssl_threads_stubs ssl crypto cstruct_stubs lwt_unix_stubs bigarray unix c"
+                # `m` and `pthreads` are built-in musl
+                static_link() {
+                    local LIBS="$* $COMMON_LIBS"
+                    echo '(-noautolink'
+                    echo ' -cclib -Wl,-Bstatic'
+                    echo ' -cclib -static-libgcc'
+                    for l in $LIBS; do
+                        echo " -cclib -l$l"
+                    done
+                    echo ')'
+                }
+                ;;
+            *)
+                echo "Please run in Alpine to avoids glibc constraints"
+                exit 2
+        esac
+        ;;
+    Darwin)
+        COMMON_LIBS="camlstr base_stubs ssl_threads_stubs /usr/local/opt/openssl/lib/libssl.a /usr/local/opt/openssl/lib/libcrypto.a cstruct_stubs lwt_unix_stubs bigarray unix pthread"
+        static_link() {
+            local LIBS="$* $COMMON_LIBS"
+            echo '(-noautolink'
+            for l in $LIBS; do
+                if [ "${l%.a}" != "${l}" ]; then echo " -cclib $l"
+                else echo " -cclib -l$l"
+                fi
+            done
+            echo ')'
+        }
+        ;;
+    *)
+        echo "Static linking not configured for your platform. See $0 to contribute."
+        exit 2
+esac
+
+static_link >src/main/linking_main.sexp laolao_stubs threads camlrun
+
+static_link >src/main/linking_client.sexp threads camlrun
+
+static_link >src/main/linking_server.sexp laolao_stubs threadsnat
+
+echo "src/main/linking_*.sexp generated; remove them to restore a standard (dynamically linked) build." >&2
```

Comment:
I wish those generated files would start with a header explaining that they are machine-generated and what their purpose is. I'm not sure that it works, but would using `; foo` for comments be valid in Dune s-exprs?

Reply:
Note: I think that it may be possible to have a better interface, by using dune support for reading environment variables: set things up so that `STATIC_BINARIES=true make` does what we expect, instead of mutating the build environment as is done here.

I think that this could be done by having a rule to build `linking.sexp` as here, with a single rule to produce `linking.sexp` (not a fallback rule and a separate script) that calls a script passing whatever context is needed, plus the content of the STATIC_BINARIES variable. If `STATIC_BINARIES` is "false", the returned sexp is empty.

Sketch:

```
(rule  
  (targets linking_client.sexp)
  (action  (with-stdout-to %{targets}
    (run scripts/linking_flags.sh "client" %{env:STATIC_BINARIES=false}))))
```

Reply:
These options are really specific and not meant for normal uses, which is why I didn't think it would be worth providing a more accessible interface: for example, it won't work for any Linux distribution other than Alpine at the moment.
`make static` will actually start a Docker container to run the build, hence not polluting your workspace.

I am interested about the env variables solution though, it does look much nicer. I couldn't find documentation about `%{env:STATIC_BINARIES=false}`, is that a feature in dune < 2 (to which we are stuck at the moment due to the old OCaml version) ?

Reply:
Yes, actually it's available in Dune 1.5.0 or so. See:
- https://github.com/ocaml/dune/issues/1506#issuecomment-774483990
- https://github.com/ocaml/dune/pull/4203